### PR TITLE
[Bugfix:TAGrading] Added cookie to remember anonymous mode

### PR DESCRIPTION
**Pull request by loadedskiffer**
### What is the current behavior?
If a TA enables anonymous mode for a gradeable and then leaves the page, when they come back anonymous mode if off again. Anonymous mode is determined in the url eg. (anon_mode=1 or anon_mode=0).
pr4_2
pr4_3

### What is the new behavior?
A cookie is set when a TA enables anonymous mode so anonymous mode will always be turned on for that gradeable until they turn it off. The url is no longer used to convey information about anonymous mode.
pr4_1

### Other information?
I tested by  messing around with the gradeables in the sample course and toggling anonymous mode on and off. It worked for every assignment.

### CODE REVIEW

File: site/app/views/grading/ElectronicGraderView.php [hunk 1]
```diff
@@ -806,6 +810,7 @@ public function detailsPage(Gradeable $gradeable, $graded_gradeables, $teamless_
             "team_gradeable_view_history" => $team_gradeable_view_history,
             "view_all" => $view_all,
             "anon_mode" => $anon_mode,
+            "anon_mode_cookie" => array_key_exists('anon_mode', $_COOKIE) ? $_COOKIE['anon_mode'] : 'off',
```
Comment (cjreed121) [inline site/app/views/grading/ElectronicGraderView.php:813]: Couldn't you use only anon_mode since that is set based off of your cookie already?

**Review by cjreed121 — CHANGES_REQUESTED** [review]

File: site/app/views/grading/ElectronicGraderView.php (hunk 1 above)
Comment (loadedskiffer) [reply site/app/views/grading/ElectronicGraderView.php:813]: The javascript needs access to the cookies value e.g const status = '{{ anon_mode_cookie }}. If I just passed anon_mode I wouldn't be able to update the cookie correctly.

File: site/app/views/grading/ElectronicGraderView.php [hunk 2]
```diff
@@ -517,7 +522,7 @@ public function detailsPage(Gradeable $gradeable, $graded_gradeables, $teamless_
             else {
                 $columns[]     = ["width" => "2%",  "title" => "",                 "function" => "index"];
                 $columns[]     = ["width" => "8%", "title" => "Section",          "function" => "section"];
-                if ($this->core->getUser()->getGroup() == User::GROUP_LIMITED_ACCESS_GRADER && $gradeable->getLimitedAccessBlind() == 2) {
+                if (($this->core->getUser()->getGroup() == User::GROUP_LIMITED_ACCESS_GRADER && $gradeable->getLimitedAccessBlind() == 2) || $anon_mode) {
```
[the remark this answers is not in the document]
Comment (loadedskiffer) [reply site/app/views/grading/ElectronicGraderView.php:525]: you're right

File: site/app/views/grading/ElectronicGraderView.php [hunk 3]
```diff
@@ -443,6 +443,11 @@ public function detailsPage(Gradeable $gradeable, $graded_gradeables, $teamless_
         if ($gradeable->isPeerGrading() && $this->core->getUser()->getGroup() == User::GROUP_STUDENT) {
             $peer = true;
         }
+        if(array_key_exists('anon_mode', $_COOKIE)){
```
[the remark this answers is not in the document]
Comment (loadedskiffer) [reply site/app/views/grading/ElectronicGraderView.php:446]: yup you're right

File: site/app/templates/grading/electronic/Details.twig [hunk 4]
```diff
@@ -687,3 +686,19 @@
 {% endmacro %}
 
 {# /Column functions #}
+
+<script type="text/javascript">
```
[the remark this answers is not in the document]
Comment (loadedskiffer) [reply site/app/templates/grading/electronic/Details.twig:690]: when I put it in details.js I get an unused function warning and then it never gets called when the button is clicked so I'm not sure.

File: site/app/views/grading/ElectronicGraderView.php (hunk 1 above)
Comment (cjreed121) [reply site/app/views/grading/ElectronicGraderView.php:813]: You could pass anon_mode into your JavaScript instead though.

File: site/app/templates/grading/electronic/Details.twig (hunk 4 above)
[the remark this answers is not in the document]
Comment (cjreed121) [reply site/app/templates/grading/electronic/Details.twig]: Missed the twig portion so this must remain in the twig file.

**Comment by cjreed121** [comment]
You should also be able to remove the $anon_mode = false from the from details function argument since that takes in from url which is no longer necessary.

File: site/app/controllers/grading/ElectronicGraderController.php [hunk 5]
```diff
@@ -811,7 +811,12 @@ public function showDetails($gradeable_id, $view = null, $sort = "id", $directio
             $this->core->addErrorMessage("You do not have permission to grade {$gradeable->getTitle()}");
             $this->core->redirect($this->core->buildCourseUrl());
         }
-
+        $anon_mode = false;
+        if (array_key_exists('anon_mode', $_COOKIE)) {
+            if ($_COOKIE['anon_mode'] === 'on') {
+                $anon_mode = true;
+            }
+        }
```
Comment (MasterOdin) [inline site/app/controllers/grading/ElectronicGraderController.php:819]: We can make this a bit more succinct:
```suggestion
        $anon_mode = isset($_COOKIE['anon_mode']) && $_COOKIE['anon_mode'] === 'on';
```

While `array_key_exists` and `isset` are mostly the same, it's better to use `isset` unless you need some specific functionality of `array_key_exists` as it's a bit faster.

**Review by cjreed121 — APPROVED** [review]

File: site/app/templates/grading/electronic/Details.twig [hunk 6]
```diff
@@ -687,3 +686,16 @@
 {% endmacro %}
 
 {# /Column functions #}
+<script>
+    function changeAnon() {
+        const status = '{{ anon_mode }}';
```
Comment (MasterOdin) [inline site/app/templates/grading/electronic/Details.twig:691]: Hm, when `anon_mode === true`, does this evaluate in twig to be `const status = '1'` and then for `anon_mode === false`, we get `const status = ''`? If possible, I think I would prefer doing something like  `const status = {{ anon_mode ? 1 : 0 }}` to make what's going on a bit less ambiguous if you don't know how PHP coerces boolean to strings.

**Review by MasterOdin — CHANGES_REQUESTED** [review]
Small comment, just want to confirm what's going on, and then I think just slight tweak and good to merge.

File: site/app/templates/grading/electronic/Details.twig (hunk 6 above)
Comment (loadedskiffer) [reply site/app/templates/grading/electronic/Details.twig:691]: Ok that makes sense, thanks

**Pull request merged**